src/Mod/PartDesign/App/FeatureSketchBased.cpp

```diff
@@ -628,12 +636,56 @@ void ProfileBased::generatePrism(TopoDS_Shape& prism,
                 throw Base::ValueError("Cannot create a pocket with a depth of zero.");
         }
 
-        // Its better not to use BRepFeat_MakePrism here even if we have a support because the
-        // resulting shape creates problems with Pocket
-        BRepPrimAPI_MakePrism PrismMaker(from, Ltotal * gp_Vec(dir), 0, 1); // finite prism
-        if (!PrismMaker.IsDone())
-            throw Base::RuntimeError("ProfileBased: Length: Could not extrude the sketch!");
-        prism = PrismMaker.Shape();
+        if (fabs(angle) > Precision::Angular() || fabs(angle2) > Precision::Angular()) {
+            // we have to make tapered prism
+
+            if (reversed)
+                dirTaper.Reverse();
+ 
+            std::list<TopoDS_Shape> drafts;
+            bool isSolid = true; // in PD we only generate solids, while Part Extrude can also create only shells
+            if (method == "ThroughAll")

```

Comment:
In order to further reduce the size of this function the determination of the parameters for the tapered prism (L, L2, angle, and angle2) could be moved to a new function.
```
void ProfileBased::determineTaperedPrismParameters(const std::string& method, bool midplane, double& L, double& L2, double& Angle, double& Angle2)
{
    if (method == "ThroughAll") {
      L = getThroughAllLength();
      L2 = 0.0;
      Angle2 = 0.0;
    }
    else if (method == "TwoLengths") {
       // nothing to do
    }
    else if (method == "Length") {
        if (midplane) {
            L = L / 2;
            L2 = L;
            Angle2 = Angle;
        }
        else {
            L2 = 0.0;
            Angle2 = 0.0;
        }
    }
}
```
And generateTaperedPrism() will become
```
void ProfileBased::generateTaperedPrism(...)
{
...
    Part::ExtrusionHelper::makeDraft(sketchshape, dirTaper, L, L2,
                                                         Base::toRadians(angle), Base::toRadians(angle2), isSolid, drafts);
    if (drafts.empty()) {
...
    }
...
}
```

Reply:
> In order to further reduce the size of this function the determination of the parameters for the tapered prism (L, L2, angle, and angle2) could be moved to a new function.

My new plan is to do this as Part Extrude does: check all parameters then store them in an enum. The enum is then input to either the draft prism or straight prism creation void.

I write new plan because since the community (forum) wants the taper feature to handle from the beginning also inner wires, I first need to merge PR #5367. This PR is from my point of view ready to be merged. I need to adapt the PR here then to the new part code.
Therefore it would help me when you could also review PR #5367.